### ISSUE TITLE
Some extra checks for validate

### ISSUE THREAD

**Issue by stefanpauliuk** [IndEcol/IEDC_tools#7]
@nheeren I now managed to dive into IEDC_tools and am very impressed about the quality and readabiliy of the code! It is a pleasure to work with and learn from it.

For the items below, I actually should have forked the repo and created a pull request for each single item... Now, the changes are already in the main branch, but we can change this if needed:

1. The ipynb tutorial notebook is very handy but outdated, as there is now a split between list and table functions etc. I took the core commands from this notebook to create my own single-file upload script: [https://github.com/IndEcol/IEDC_tools/blob/master/IEDC_upload_single.py](https://github.com/IndEcol/IEDC_tools/blob/master/IEDC_upload_single.py)
2. I added additional comments and descriptions to some of the functions that I used.
3. In _check_datasets_entry_ I modified the code so that the function behaves as expected when the create/update/replace flags are set or not.
4. In _upload_data_table_, I added the existence check for the (dataset_name,dataset_version) tuple, and this function now only uploads data if the dataset entry is already present in the dataset table but no data for this dataset have been uploaded to the data table yet. Error messages are issued accordingly.
5. (that took the longest time to figure out) In the possible case that there are duplicate entries for a _oto_ aspect (which is possible as they can be empty, see example here: IndEcol/IE_data_commons#25), the following command will produce a table/df longer than the original one, with a duplicate match for each affected item: `tmp = data.merge(db_classitems2, left_on=class_name, right_on='attribute%s_oto' % str(int(attribute_no)), how='left')` both list and table upload are affected. My proposed solution is a new check in the upload routine to report and avoid such double instances: `if len(tmp.index) != len(data.index): raise AssertionError("The database classification table contains at least one conflicting duplicate entry for the unique attribute attribute%s_oto of classification %s. Data upload halted. Check classification for duplicate entries!" % (str(int(attribute_no)), class_name))` This way, the upload routine performs additional consistency checks on the db, which can avoid errors that have been overlooked so far.

**Comment by nheeren** [comment]
@stefanpauliuk Wow -- great you were able to comprehend and run everything as there is room for better documentation (as always ;) Please let me know if you want to discuss anything or would like me to look into specific items. Developing this whole thing in our close collaboration was actually a lot of fun.

Your changes are great and the single upload file is a good idea. Probably we should merge that with the debug_xx.py files one day. Also, we should add .pyc files to .gitignore as they are part of your commit now. 

As far as I understand there is no action required here and we can close this issue, right?

**Comment by stefanpauliuk** [comment]
Thanks for the quick reply @nheeren !
All is good for the moment :)

- I added the .pyc filed to .gitignore and a whole bunch of other stuff (just copied my default .gitignore appendix below the shielded PW files)
- I managed to fix the Kosovo-Trinidad issue in the IEDC. ca. 40 datasets, all from the global steel cycle project data group, where affected but this is fixed now.
- Agree that we can consider merging the single upload with the debug_xx.py files in principle, but let's first see what the use cases are in the future!
This is very exciting and I am happy that this project is still developing. Many good possibilities ahead!
